locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/softlayer/SoftLayerSameVlanLocationCustomizer.java

```diff
@@ -0,0 +1,398 @@
+
+    /**
+     * Update the {@link org.jclouds.compute.options.TemplateOptions} that will
+     * be used by {@link JcloudsLocation} to obtain machines. Uses the VLAN
+     * numbers configured on an existing machine, as saved in the configuration
+     * maps for {@link #PUBLIC_VLAN_ID_MAP public} and {@link #PRIVATE_VLAN_ID_MAP private}
+     * VLAN numbers.
+     *
+     * @see {@link JcloudsLocationCustomizer#customize(JcloudsLocation, ComputeService, TemplateOptions)}
+     */
+    @Override
+    public void customize(JcloudsLocation location, ComputeService computeService, TemplateOptions templateOptions) {
+        // Check we are customising a SoftLayer location
+        String provider = location.getProvider();
+        if (!(provider.equals(SoftLayerConstants.SOFTLAYER_PROVIDER_NAME) &&
+                templateOptions instanceof SoftLayerTemplateOptions)) {
+            String message = String.format("Invalid location provider or template options: %s/%s",
+                    provider, templateOptions.getClass().getSimpleName());
+            LOG.warn(message);
+            throw new IllegalArgumentException(message);
+        }
+
+        // Check template options for VLAN configuration and save them if set
+        String scopeUid = getScopeUid(location);
+        SoftLayerTemplateOptions softLayerOptions = (SoftLayerTemplateOptions) templateOptions;
+        Integer publicVlanId = softLayerOptions.getPrimaryNetworkComponentNetworkVlanId();
+        Integer privateVlanId = softLayerOptions.getPrimaryBackendNetworkComponentNetworkVlanId();
+        if (publicVlanId != null || privateVlanId != null) {
+            LOG.debug("SoftLayer VLANs private {} and public {} already configured in template options for scope: {}",
+                    new Object[] { privateVlanId, publicVlanId, scopeUid });
+            savePublicVlanId(location, scopeUid, publicVlanId);
+            savePrivateVlanId(location, scopeUid, privateVlanId);
+            return;
+        }
+
+        // Create a new latch if we are first
+        CountDownLatch latch = null;
+        synchronized (lock) {
+            latch = lookupCountDownLatch(location, scopeUid);
+            if (latch == null) {
+                // We are first to try obtaining a VM; create a latch and return
+                LOG.debug("Creating new latch for scope: {}", scopeUid);
+                latch = createCountDownLatch(location, scopeUid);
+                return;
+            }
+        }
+
+        // Try and acquire the latch, or time out
+        Duration timeout = getTimeout(location);
+        Tasks.setBlockingDetails("Waiting for VLAN details");
+        try {
+            LOG.debug("Waiting for VLAN details for scope: {}", scopeUid);
+            if (!Uninterruptibles.awaitUninterruptibly(latch, timeout.toMilliseconds(), TimeUnit.MILLISECONDS)) {
+                // Release the latch and throw an exception to prevent others blocking forever
+                latch.countDown();
```

Comment:
confused here -- imagine provisioning threads A B C and D started, in that order.  i think then:

* A keeps the latch for longer than timeout
* B detects this, clears the latches, and throws
* C picks up the latch
* D is very soon going to timeout and repeat B's behaviour

Feels like B,C, and D need symmetry, and their timers reset if there is a new master, and A should be the thread to fail, rather than B (who here has done nothing wrong apart from being the whistleblower).

I was thinking `SemaphoreWithOwner` because then A can check whether the owner has been stolen from him and fail, B can claim ownership, and C & D can check whether the owner has changed and reset their timer appropriately.  (I think it's okay if C & D wait for the initial timeout to expire before detecting and resetting, btw; it's quite a bit more work to get them to reset their timers as soon as B steals ownership and not necessary IMO.)